Docs: How to organize data - Group path (#4126)
aiidateam/aiida-core: docs/source/howto/data.rst

Organizing data
===============

.. _how-to:data:organize:grouppath:

Organise groups in hierarchies
------------------------------

.. meta::
   :description: The reStructuredText plaintext markup language
   :keywords: grouppath

Groups in AiiDA are inherently "flat", in that groups may only contain nodes and not other groups.

> However, the `GroupPath` utility allows one to construct *virtual* group hierarchies based on delimited group labels.

Reviewer: ```suggestion
However it is possible to construct *virtual* group hierarchies based on delimited group labels using the `GroupPath` utility.
```

Reviewer: Can you make sure that `GroupPath` links to the API reference?

> `GroupPath` is designed to work in much the same way as python's `pathlib.Path`, whereby paths are denoted by forward slash characters '/' in group labels.

Reviewer: ```suggestion
`GroupPath` is designed to work in much the same way as Python's `pathlib.Path`, whereby paths are denoted by forward slash characters '/' in group labels.
```

Reviewer: Can you link both `GroupPath` and `pahtlib.Path` to the respective API reference docs?

You can also traverse directly through the nodes of a path, optionally filtering by node class and any other filters allowed by the `QueryBuilder`:

.. code-block:: ipython

    In [17]: from aiida.orm import Data
    In [18]: data = Data()
    In [19]: data.set_extra("key", "value")
    In [20]: data.store()
    Out[20]: <Data: uuid: 0adb5224-585d-4fd4-99ae-20a071972ddd (pk: 1)>
    In [21]: path["base1/sub_group1"].get_group().add_nodes(data)
    In [21]: next(path.walk_nodes(node_class=Data, filters={"extras.key": "value"}))
    Out[21]: WalkNodeResult(group_path=GroupPath('base1/sub_group1', cls='<class 'aiida.orm.groups.Group'>'),
    node=<Data: uuid: 0adb5224-585d-4fd4-99ae-20a071972ddd (pk: 1)>)

Finally, you can also specify the `Group` subclasses (as discussed above):

.. code-block:: ipython

    In [22]: from aiida.orm import UpfFamily
    In [23]: path2 = GroupPath(cls=UpfFamily)
    In [24]: path2["base1"].get_or_create_group()
    Out[24]: (<UpfFamily: "base1" [type core.upf], of user user@email.com>, True)

.. important::

    A `GroupPath` instance will only recognise groups of the instantiated ``cls`` type.

>     The default ``cls`` is ``aiida.orm.Group``:

Reviewer: Can you convert that into an API reference?

Author: fixed in b5f6df7
